BUG: fix convert_dtypes to handle empty df

owenlamont opened the pull request:
Added a conditional to check a DataFrame had at least one column and if not return the original (empty) DataFrame if so.

- [x] closes #40393
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

owenlamont committed:
BUG 40393 convert_dtypes handle empty df

Added a conditional to check a DataFrame had at least one column and if
not return the original (empty) DataFrame if so.

owenlamont committed:
BUG: 40393 convert_dtypes allow empty df

Added unit test and added reference to whatsnew

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -6322,7 +6322,10 @@ def convert_dtypes(
+            if len(results) > 0:
+                return concat(results, axis=1, copy=False)
+            else:
+                return self
```
Since it's empty it might not matter that much, but to be sure let's return `self.copy()` (the other path through concat will also always return a new object)

jorisvandenbossche commented on pandas/tests/series/methods/test_convert_dtypes.py:
```diff
@@ -225,3 +225,10 @@ def test_convert_bool_dtype(self):
+    def test_convert_empty(self):
+        # Empty DataFrame can pass convert_dtypes, see GH#40393
+        df = pd.DataFrame().convert_dtypes()
+        assert df.empty
```
There is a similar file in the `pandas/tests/frame/methods` directory, which is probably the more suitable place to put this test

owenlamont replied:
Thanks - yeah I was a bit uncertain where best to leave that.

owenlamont replied:
Oops I see - I totally didn't notice there was separate folder hierarchies for series and frame tests. Thanks for correcting me.

jorisvandenbossche commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -625,6 +625,7 @@ Other
 - Bug in :class:`Styler` where multiple elements in CSS-selectors were not correctly added to ``table_styles`` (:issue:`39942`)
 - Bug in :meth:`DataFrame.equals`, :meth:`Series.equals`, :meth:`Index.equals` with object-dtype containing ``np.datetime64("NaT")`` or ``np.timedelta64("NaT")`` (:issue:`39650`)
 - Bug in :func:`pandas.util.show_versions` where console JSON output was not proper JSON (:issue:`39701`)
+- Bug in :meth:`generic.convert_dtypes` when called on an empty DataFrame (:issue:`40393`)
```
```suggestion
- Bug in :meth:`DataFrame.convert_dtypes` when called on an empty DataFrame (:issue:`40393`)
```

(generally we try to use the public name in the release notes)

jorisvandenbossche reviewed (commented):
Thanks for the PR! 
Just a few small comments

owenlamont committed:
BUG: 40393 convert_dtypes to handle empty df

Moved a frame test incorrectly placed in the
series hierarchy. Fixed some naming in bug
documentation. Changed convert_dtypes to
return a copy of apparently empty dataframes

owenlamont commented:
Thanks @jorisvandenbossche

All good call outs (particularly the test in the wrong place - I should've noticed that). I hope it looks better now.

rhshadrach commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -625,6 +625,7 @@ Other
 - Bug in :class:`Styler` where multiple elements in CSS-selectors were not correctly added to ``table_styles`` (:issue:`39942`)
 - Bug in :meth:`DataFrame.equals`, :meth:`Series.equals`, :meth:`Index.equals` with object-dtype containing ``np.datetime64("NaT")`` or ``np.timedelta64("NaT")`` (:issue:`39650`)
 - Bug in :func:`pandas.util.show_versions` where console JSON output was not proper JSON (:issue:`39701`)
+- Bug in :meth:`DataFrame.convert_dtypes` when called on an empty DataFrame (:issue:`40393`)
```
Can you say what the bug is, e.g. "incorrectly raised" before "when called".

owenlamont replied:
No problem - I mentioned the exception as that seemed to read better to me. Hope it looks okay to you.

rhshadrach commented on pandas/tests/frame/methods/test_convert_dtypes.py:
```diff
@@ -26,3 +26,8 @@ def test_convert_dtypes(self, convert_integer, expected):
+    def test_convert_empty(self):
+        # Empty DataFrame can pass convert_dtypes, see GH#40393
+        df = pd.DataFrame().convert_dtypes()
+        assert df.empty
```
I think it'd be better to use tm.assert_frame_equal, e.g. testing dtypes are the same.

owenlamont replied:
Done, yes I noticed I noticed the pandas testing asserts seemed to be used exclusively. I wasn't quite sure which way to use them but hopefully this change looks okay.

rhshadrach reviewed (changes requested):
Looks really good - two minor requests

owenlamont committed:
BUG: 40393 convert_dtypes handle empty df

Changed test to use tm.assert_equal and documented
error more explicitly

owenlamont commented:
Thanks for the feedback @rhshadrach I'm still not sure I've got the assert written the way you'd like - didn't know whether to compare the DataFrame to itself or to a new empty DataFrame. Please let me know if you'd like anything else updated.

owenlamont committed:
BUG: 40393 convert_dtypes allow empty df

Replaced tm.assert_equals with tm.assert_frame_equals

jorisvandenbossche commented:
Thanks a lot @owenlamont for your first contribution!

owenlamont commented:
Thank you! I'm honoured to contribute to such a foundational part of the PyData ecosystem. First of many hopefully.
